src/bootstrap/src/utils/build_stamp.rs

```diff
@@ -0,0 +1,202 @@
+//! Module for managing build stamp files.
+//!
+//! Contains the core implementation of how bootstrap utilizes stamp files on build processes.
+
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+use sha2::digest::Digest;
+
+use crate::core::builder::Builder;
+use crate::core::config::TargetSelection;
+use crate::utils::helpers::{hex_encode, mtime};
+use crate::{Compiler, Mode, helpers, t};
+
+#[cfg(test)]
+mod tests;
+
+/// Manages a stamp file to track build state. The file is created in the given
+/// directory and can have custom content and name.
+#[derive(Clone)]
+pub struct BuildStamp {
+    path: PathBuf,
+    stamp: String,
+}
+
+impl BuildStamp {
+    /// Creates a new `BuildStamp` for a given directory.
+    ///
+    /// By default, stamp will be an empty file named `.stamp` within the specified directory.
+    pub fn new(dir: &Path) -> Self {
+        // Avoid using `is_dir()` as the directory may not exist yet.
+        // It is more appropriate to assert that the path is not a file.
+        assert!(!dir.is_file(), "can't be a file path");
+        Self { path: dir.join(".stamp"), stamp: String::new() }
+    }
+
+    /// Returns path of the stamp file.
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    /// Returns the value of the stamp.
+    ///
+    /// Note that this is empty by default and is populated using `BuildStamp::add_stamp`.
+    /// It is not read from an actual file, but rather it holds the value that will be used
+    /// when `BuildStamp::write` is called.
+    pub fn stamp(&self) -> &str {
+        &self.stamp
+    }
+
+    /// Sets stamp content to the specified value.
```

Comment:
Nit: add `stamp` to stamp content?

Reply:
Oops, yeah the comment should be updated too here since the last update.